rust_snuba/ARCHITECTURE.md

```diff
@@ -0,0 +1,101 @@
+# rust-snuba consumers
+
+Snuba has two ways to ingest data into clickhouse: The old, Python-based consumers, and this Rust consumer.
+
+This document is about the Rust one, but due to it starting out as a
+line-by-line port of the Python consumer, the higher-level architecture also
+applies to Python.
+
+## Entrypoint
+
+`rust_snuba` is a library crate that is being compiled with PyO3/maturin into a
+Python wheel. This Python wheel exposes a `consumer` function which is used
+from within the Snuba (Python) CLI to launch `snuba rust-consumer`.
+Configuration parsing, CLI parsing and runtime config are still handled in
+Python so as to avoid unnecessary code duplication and potentially multiple
+sources of truth for config.
+
+## Steps
+
+`rust_snuba` then uses the Rust port of Arroyo to define its own consumer. The following steps (processing strategies) exist (see `factory.rs`):
+
+1. `HealthCheck` step, touches a file in irregular intervals to signal to k8s
+   that the consumer is still alive.
+2. `SetJoinTimeout` step, resets the join timeout to 0 for the next steps. This
+   marks a section where it's ok to drop in-flight progress during rebalancing.
+3. Message transformers, basically a threadpool to parse messages, validate
+   them and turn them into rows. This can call both back into Python code for
+   when the MessageProcessor hasn't been ported to Rust yet, but most datasets
+   have a Rust-native processor.
+4. `Reduce` opens batches by opening a connection to clickhouse, and writes
+   rows into them one-by-one. There is only one batch at a time, and the size
+   of a batch is constrained by `--max-batch-time-ms` and `--max-batch-size`.
+5. `SetJoinTimeout` is reset to "no timeout" here. Between this and the
+   previous section, everything will be dropped by Arroyo during rebalancing as
+   parsing and validating messages has no side effects. Even sending data to
+   ClickHouse is idempotent for as long as the batch is not closed.
+5. `ClickhouseWriterStep` is responsible for closing batches. It takes a batch
+   from the previous step, signals to ClickHouse that the batch is done, and
+   waits for ClickHouse to return 200 OK.
+6. COGS and commit log are two topics we produce metadata into once we're done
+   writing to ClickHouse, but they are typically not relevant for performance.
```

Comment:
Just curious, what's the intuition for figuring out that this is not relevant for performance?